### PR TITLE
[ST-704] add cypress test runner

### PR DESCRIPTION
**Pull request by Konboi**
cypress JUnit reporter outputs incorrect format(https://github.com/cypress-io/cypress/issues/8237), so we can't use file command.
I create a cypress plugin for that.

I need some testsuites elements to get the test file name that use to report, so I define testsuites method that can overwrite.
And I prepare custom testsuites method for cypress.

outputed XML: https://github.com/launchableinc/cli/pull/101/files#diff-57f8b2c67ad176f3307742a51acc476264050670ebe13c6492441fd17596f971

### CODE REVIEW

File: launchable/test_runners/cypress.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    for r in reports:
+        client.report(r)
+
+    def testsuites(xml) -> [TestSuite]:
+        testsuites = []
+        if isinstance(xml, JUnitXml):
+            filepath = xml._elem.find(
+                './/testsuite[@name="Root Suite"]').get("file")
+            for suite in xml:
+                suite._elem.attrib.update(
```
Comment (Konboi) [inline launchable/test_runners/cypress.py:20]: here is the main part of a cypress plugin. 
Get a `file` attribute value used by XPath and add it to a testsuite attribute for each testsuites, likes this (based on https://github.com/launchableinc/cli/pull/101/files#diff-57f8b2c67ad176f3307742a51acc476264050670ebe13c6492441fd17596f971).

```xml
<?xml version="1.0" encoding="UTF-8"?>
<testsuites name="Mocha Tests" time="1.1780" tests="6" failures="0">
  <testsuite name="Root Suite" timestamp="2021-01-20T08:49:02" tests="0" file="cypress/integration/examples/window.spec.js" time="0.0000" failures="0">
  </testsuite>
  <testsuite name="Window" timestamp="2021-01-20T08:49:02" tests="3" time="0.3150" failures="0">
    ...
  </testsuite>
</testsuites> 
```
↓
```xml
<?xml version="1.0" encoding="UTF-8"?>
<testsuites name="Mocha Tests" time="1.1780" tests="6" failures="0">
  <testsuite name="Root Suite" timestamp="2021-01-20T08:49:02" tests="0" file="cypress/integration/examples/window.spec.js" time="0.0000" failures="0" filepath="cypress/integration/examples/window.spec.js">
  </testsuite>
  <testsuite name="Window" timestamp="2021-01-20T08:49:02" tests="3" time="0.3150" failures="0" filepath="cypress/integration/examples/window.spec.js">
    ...
  </testsuite>
</testsuites> 
```

But, I'm may sure there's a better way...

**Review by Konboi — COMMENTED** [review]
I'm going to write formal documents, so I write only simple usage for now.

```
launchable record tests --session <SESSION> cypress <REPROT FILES>
cypress run --spec "$(find **/*.spec.js |  launchable subset --target 10% --session <SESSION> cypress)"
```

**Comment by kohsuke** [comment]
I'm saying this without fully understanding this change, but ...

The thing that gives me a pause is the tweak to `RecordTests`. The `testsuites` indirection wasn't obvious to me. It is ostensibly to map JUnit-esque objects to a list of `TestSuite`, but it's actually used to manipulate the parsed JUnit report before it gets processed. That use case is understandable, and for that it's more straightforward to provide an indirection around `JUnitXml.fromfile(p)`.

**Review by kohsuke — APPROVED** [review]
With that said, I can see that this works, so if you decide this is good for now that's fine by me.

**Comment by Konboi** [comment]
> The testsuites indirection wasn't obvious to me. It is ostensibly to map JUnit-esque objects to a list of TestSuite, but it's actually used to manipulate the parsed JUnit report before it gets processed.

Yes, exactly true. I think this method's name and way aren't the best too.

> That use case is understandable, and for that it's more straightforward to provide an indirection around JUnitXml.fromfile(p).

you mean,  we use `JUnitXML.fromfile` method basically but when we have to parse invalid formats, we provide a custom method that returns was parsed, and formatted XML elements object and override the method?



**Comment by ninjinkun** [comment]
You modified (or more precisely "normalized") the XML file in walking test suites processing processing process. IMO,if you need to modify the XML file you can write a converter such as `normalized_xml = some_converter(JUnitXML.fromFile(p))`.

File: launchable/commands/record/tests.py [hunk 2]
```diff
@@ -41,6 +42,22 @@ def tests(context, base_path, session_id: str):
...
+
+        @staticmethod
+        def default_testsuites(xml) -> [TestSuite]:
+            if isinstance(xml, JUnitXml):
+                testsuites = [suite for suite in xml]
+            elif isinstance(xml, TestSuite):
+                testsuites = [xml]
+            else:
+                # TODO: what is a Pythonesque way to do this?
+                assert False
```
[the remark this answers is not in the document]
Comment (ninjinkun) [reply launchable/commands/record/tests.py:57]: But according to the Launchable CLI design principle, you should branch an exception line and suppress the error line with LAUNCHABLE_REPORT_ERROR variable.

**Comment by kohsuke** [comment]
> > That use case is understandable, and for that it's more straightforward to provide an indirection around JUnitXml.fromfile(p).
> 
> you mean, we use `JUnitXML.fromfile` method basically but when we have to parse invalid formats, we provide a custom method that returns was parsed, and formatted XML elements object and override the method?

Exactly!

**Review by Konboi — COMMENTED** [review]
I tried to fix it to use by junitparser.JUnitXml.fromfile parse_func option.
Will you please review again? @ninjinkun

**Review by ninjinkun — APPROVED** [review]
Beautiful solution! I love it.

**Pull request merged**